### PR TITLE
Closes #5115 - New Preload: thirdpary classes

### LINKED ISSUE #5115: New Preload: Add thirdparties classes
Thirdparty classes from:
- Yoast
- All in 1 SEO
- Jetpack
- SEOPress
- Math Rank
are obsolete and needs to be refactored.
Only Jetpack needs to send a sitemap URL with the filter `rocket_sitemap_preload_list` as there is no redirection from the Wordpress sitemap to their.

### PR DESCRIPTION
**Pull request by CrochetFeve0251**
## Description

Refactor all Thirdparties linked too the preload in the old architecture to the new one.
For that theses thirdparties got refactored:
- SEOPress.
- Jetpack.
- Rank Math SEO
- All in One SEO.

Fixes #5115

## Type of change

- [ ] Enhancement (non-breaking change which improves an existing functionality)

## How Has This Been Tested

- [ ] Automated tests

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: inc/ThirdParty/Plugins/AllInOneSEOPack.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	 * @return Array Updated array of sitemaps to preload
+	 */
+	public function add_all_in_one_seo_sitemap( $sitemaps ) {
+		if ( ! $this->option->get( 'all_in_one_seo_xml_sitemap', false ) ) {
+			return $sitemaps;
+		}
+
+		$aioseo_v3 = defined( 'AIOSEOP_VERSION' );
+		$aioseo_v4 = defined( 'AIOSEO_VERSION' ) && function_exists( 'aioseo' );
+		if ( ! $aioseo_v3 && ! $aioseo_v4 ) {
```
Comment (jeawhanlee) [inline inc/ThirdParty/Plugins/AllInOneSEOPack.php]: Do we need to do this check here again, since we already have one on the subscribed events method.

**Review by jeawhanlee — COMMENTED** [review]
I think we should have all the SEO Related thirdparty classes in the SEO folder and also update namespaces.
- [AllInOneSEOPack](https://github.com/wp-media/wp-rocket/blob/task/preload-thirdpary-classes/inc/ThirdParty/Plugins/AllInOneSEOPack.php)
- [Jetpack.php](https://github.com/wp-media/wp-rocket/blob/task/preload-thirdpary-classes/inc/ThirdParty/Plugins/Jetpack.php)
- [RankMathSEO.php](https://github.com/wp-media/wp-rocket/blob/task/preload-thirdpary-classes/inc/ThirdParty/Plugins/RankMathSEO.php)

File: inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	/**
+	 * Add All in One SEO Sitemap to the preload list
+	 *
+	 * @param Array $sitemaps Array of sitemaps to preload.
+	 * @return Array Updated array of sitemaps to preload
+	 */
+	public function add_all_in_one_seo_sitemap( $sitemaps ) {
+		if ( ! $this->option->get( 'all_in_one_seo_xml_sitemap', false ) ) {
+			return $sitemaps;
+		}
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php]:
```suggestion
```
Option won't exist anymore with the new system

File: inc/ThirdParty/Plugins/SEO/RankMathSEO.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	 * @param array $sitemaps Sitemaps to preload.
+	 * @return array Updated Sitemaps to preload
+	 */
+	public function rocket_sitemap( $sitemaps ) { // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals
+		if ( $this->option->get( 'rank_math_xml_sitemap', false ) ) {
+			$sitemaps[] = Router::get_base_url( 'sitemap_index.xml' );
+		}
+
+		return $sitemaps;
+	}
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/SEO/RankMathSEO.php]:
```suggestion
	public function add_sitemap_preload( $sitemaps ) { // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals
		$sitemaps[] = Router::get_base_url( 'sitemap_index.xml' );

		return $sitemaps;
	}
```

File: inc/ThirdParty/Plugins/SEO/SEOPress.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	/**
+	 * Add SEOPress sitemap URL to the sitemaps to preload
+	 *
+	 * @param array $sitemaps Sitemaps to preload.
+	 * @return array Updated Sitemaps to preload
+	 */
+	public function add_seopress_sitemap( $sitemaps ) {
+		if ( $this->option->get( 'seopress_xml_sitemap', false ) ) {
+			$sitemaps[] = get_home_url() . '/sitemaps.xml';
+		}
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/SEO/SEOPress.php]:
```suggestion
		$sitemaps[] = get_home_url() . '/sitemaps.xml';
```

File: inc/ThirdParty/Plugins/SEO/TheSEOFramework.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+	/**
+	 * Adds TSF sitemap URLs to preload.
+	 *
+	 * @param array $sitemaps Sitemaps to preload.
+	 * @return array Updated Sitemaps to preload
+	 */
+	public function add_tsf_sitemap_to_preload( $sitemaps ) {
+
+		if ( $this->option->get( 'tsf_xml_sitemap', false ) ) {
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/SEO/TheSEOFramework.php]:
```suggestion
```

File: inc/ThirdParty/Plugins/SEO/TheSEOFramework.php [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+					// In other words, this enables support for ALL current and future public sitemap endpoints.
+					if ( ! empty( $data['robots'] ) ) {
+						$sitemaps[] = $sitemap_bridge->get_expected_sitemap_endpoint_url( $id );
+					}
+				}
+			} else {
+				// Deprecated. TSF <4.0.
+				$sitemaps[] = the_seo_framework()->get_sitemap_xml_url();
+			}
+		}
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/SEO/TheSEOFramework.php:82]:
```suggestion
```

**Review by remyperona — CHANGES_REQUESTED** [review]
On the jetpack file, only keep the method that is adding the sitemap URL to the preload. All the rest is either not necessary anymore, or not working correctly. The previous compatibility file wasn't ever loaded because of this

File: inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+class AllInOneSEOPack implements Subscriber_Interface {
+
+	/**
+	 * Option instance.
+	 *
+	 * @var Options_Data
+	 */
+	protected $option;
+
```
Comment (jeawhanlee) [inline inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php:16]:
```suggestion
/**
     * All in one pack version
     *
     * @var bool
     */
    private $aioseo_v3;

    /**
     * All in one pack version
     *
     * @var bool
     */
    private $aioseo_v4;
```

File: inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+	/**
+	 * Instantiate class.
+	 *
+	 * @param Options_Data $option Option instance.
+	 */
+	public function __construct( Options_Data $option ) {
+		$this->option = $option;
```
Comment (jeawhanlee) [inline inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php:23]:
```suggestion
		$this->option = $option;
		$this->aioseo_v3 = defined( 'AIOSEOP_VERSION' );
        $this->aioseo_v4 = defined( 'AIOSEO_VERSION' ) && function_exists( 'aioseo' );
```

File: inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	/**
+	 * Subscribed events.
+	 */
+	public static function get_subscribed_events() {
+		$aioseo_v3 = defined( 'AIOSEOP_VERSION' );
+		$aioseo_v4 = defined( 'AIOSEO_VERSION' ) && function_exists( 'aioseo' );
+
+		if ( ! $aioseo_v3 && ! $aioseo_v4 ) {
```
Comment (jeawhanlee) [inline inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php:33]:
```suggestion
		if ( ! $this->aioseo_v3 && ! $this->aioseo_v4 ) {
```

File: inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+	/**
+	 * Add All in One SEO Sitemap to the preload list
+	 *
+	 * @param Array $sitemaps Array of sitemaps to preload.
+	 * @return Array Updated array of sitemaps to preload
+	 */
+	public function add_all_in_one_seo_sitemap( $sitemaps ) {
+
+		$aioseo_v3 = defined( 'AIOSEOP_VERSION' );
+		$aioseo_v4 = defined( 'AIOSEO_VERSION' ) && function_exists( 'aioseo' );
```
Comment (jeawhanlee) [inline inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php:51]:
```suggestion
```

File: inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+		$sitemap_enabled = false;
+		if ( $aioseo_v3 && ! $aioseo_v4 ) {
```
Comment (jeawhanlee) [inline inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php:54]:
```suggestion
		if ( $this->aioseo_v3 && ! $this->aioseo_v4 ) {
```

File: inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+			$aioseop_options = get_option( 'aioseop_options' );
+			$sitemap_enabled = ( isset( $aioseop_options['modules']['aiosp_feature_manager_options']['aiosp_feature_manager_enable_sitemap'] ) && 'on' === $aioseop_options['modules']['aiosp_feature_manager_options']['aiosp_feature_manager_enable_sitemap'] ) || ( ! isset( $aioseop_options['modules']['aiosp_feature_manager_options'] ) && isset( $aioseop_options['modules']['aiosp_sitemap_options'] ) );
+		}
+
+		if (
+			( ! $aioseo_v4 && ! $sitemap_enabled ) ||
+			( $aioseo_v4 && ! aioseo()->options->sitemap->general->enable )
+		) {
```
Comment (jeawhanlee) [inline inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php:62]:
```suggestion
		if (
            ( ! $this->aioseo_v4 && ! $sitemap_enabled ) ||
            ( $this->aioseo_v4 && ! aioseo()->options->sitemap->general->enable )
        ) {
```

File: inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+			return $sitemaps;
+		}
+
+		if ( $aioseo_v3 ) {
```
Comment (jeawhanlee) [inline inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php:66]:
```suggestion
		if ( $this->aioseo_v3 ) {
```

File: inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+			$sitemaps[] = trailingslashit( home_url() ) . apply_filters( 'aiosp_sitemap_filename', 'sitemap' ) . '.xml'; // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals
+		} elseif ( $aioseo_v4 ) {
```
Comment (jeawhanlee) [inline inc/ThirdParty/Plugins/SEO/AllInOneSEOPack.php:68]:
```suggestion
		} elseif ( $this->aioseo_v4 ) {
```

**Review by jeawhanlee — COMMENTED** [review]
What do you think about this?

**Comment by CrochetFeve0251** [comment]
> What do you think about this?

@jeawhanlee 
From the state of the code and the architecture that would be nice. 
However there are two problems to that:
- I have no idea how to test it with integration tests if we move the check in the constructor as we need to tests both behaviors and they are depending on if the function is existing or not.
- The method to subscribe event is static so we cannot call $this inside it.

**Comment by jeawhanlee** [comment]
True - the static method.

**Review by remyperona — CHANGES_REQUESTED** [review]
Also remove the cookie law related code from jetpack

**Comment by CrochetFeve0251** [comment]
> Also remove the cookie law related code from jetpack

Oh ok left it as it had nothing to do with sitemap

**Review by remyperona — APPROVED** [review]

**Pull request merged**